clients/python-wrapper/lakefs/object_io.py

```diff
@@ -0,0 +1,163 @@
+
+    def exists(self):
+        try:
+            self._client.sdk_client.objects_api.head_object(self._repo, self._ref, self._path)
+            return True
+        except NotFoundException:
+            return False
+
+
+class WriteableObject(ReadableObject):
+    """
+    WriteableObject inherits from ReadableObject and provides read/write functionality for lakeFS objects 
+    using IO semantics.
+    This Object is instantiated and returned upon invoking open() on Branch reference type.
+    """
+
+    def __init__(self, repository: str, reference: str, path: str,
+                 pre_sign: Optional[bool] = None, client: Optional[Client] = DefaultClient) -> None:
+        # Verify that reference is a branch, otherwise throws exception
+        try:
+            client.sdk_client.branches_api.get_branch(repository, reference)
+        except NotFoundException:
+            raise UnsupportedOperationException("reference is not an existing branch")
+
+        super().__init__(repository, reference, path, pre_sign, client=client)
+
+    def create(self,
+               data: UploadContentType,
+               mode: Literal['x', 'xb', 'w', 'wb'] = 'wb',
+               pre_sign: Optional[bool] = None,
+               content_type: Optional[str] = None,
+               metadata: Optional[dict[str, str]] = None):
+        content = data
+        if mode.startswith('x') and self.exists():  # Requires explicit create
+            raise ObjectExistsException
+
+        binary_mode = mode.endswith('b')
+        if binary_mode and isinstance(data, str):
+            content = data.encode('utf-8')
+        elif not binary_mode and isinstance(data, bytes):
+            content = data.decode('utf-8')
+        # TODO: handle streams
+        is_presign = pre_sign if pre_sign is not None else self._pre_sign
+        stats = self._upload(content, is_presign, content_type, metadata)
+        # reset pos after create
+        self._pos = 0
+        return stats
+
+    def delete(self):
+        # TODO: Implement
+        raise NotImplementedError
+
+    @staticmethod
+    def _extract_etag_from_response(headers) -> str:
+        # prefer Content-MD5 if exists
+        content_md5 = headers.get("Content-MD5")
+        if content_md5 is not None and len(content_md5) > 0:
+            try:  # decode base64, return as hex
+                decode_md5 = base64.b64decode(content_md5)
+                return binascii.hexlify(decode_md5).decode("utf-8")
+            except binascii.Error:
+                pass
+
+        # fallback to ETag
+        etag = headers.get("ETag", "").strip(' "')
+        return etag
+
+    def _upload(self, content, pre_sign, content_type: Optional[str] = None,
+                metadata: Optional[dict[str, str]] = None) -> ObjectStats:
+        if not pre_sign:
+            raise UnsupportedOperationException("Upload currently supported only in pre-sign mode")
```

Comment:
Assuming this will be added soon?

Reply:
As soon as possible

Reply:
https://github.com/treeverse/lakeFS/issues/6963